### PR TITLE
Event schema

### PR DESCRIPTION
**Pull request by d-chambers**
This PR adds an event schema using [pydantic](https://pydantic-docs.helpmanual.io/) then uses pydantic to to perform the conversions of obspy.Catalog objects to and from json/dicts.

This is primarily an internal change and shouldn't really effect ObsPlus' external API.

### CODE REVIEW

**Comment by d-chambers** [comment]
I should mention I had to slightly modify test catalog 4 (tests/test_data/test_catalogs/cat4.xml) because ObsPy's quakeml parser expected ints for `CompositeTime`'s year, month, day, hour, and minute uncertainties. However, this is inconsistent with `QuantityError`specifying its types as floats. Since this is such an obscure corner case (I have never seen a `CompositeTime` in the wild) I just removed the uncertainties from the test catalog.

**Comment by d-chambers** [comment]
I ended up finding a bug where `WaveBank` couldn't be pickled because the index cached used itertools.cycle. I replaced this with a simpler incrementing index attribute.

**Comment by d-chambers** [comment]
Anyone have a sec to review this? I am not too worried about it since the external APIs haven't changed but another pair of eyes couldn't hurt.

File: obsplus/events/json.py [hunk 1]
```diff
@@ -6,136 +6,44 @@
...
+import obsplus.events.schema as event_schema
 
 JSON_SERIALIZER_VERSION = "0.0.0"  # increment when serialization changes
 
 
 # -------------------- events to json functions
 
 
-def cat_to_json(catalog: obspy.Catalog) -> str:
+def cat_to_json(events: Union[Catalog, Event]) -> str:
```
Comment (calum-chamberlain) [inline obsplus/events/json.py:19]: Good to see single events explicitly allowed, but would it be possible to keep the variable as `catalog`? I'm pretty sure I use this function directly in a few of my codes, and I'm concerned that I may have used a kwarg at some point.

File: obsplus/events/json.py [hunk 2]
```diff
@@ -6,136 +6,44 @@
...
+    if isinstance(events, Event):  # A single event was passed
+        events = Catalog(events=[events])
+    elif not isinstance(events, Catalog):  # sequence was passed
+        events = Catalog(events=events)
+    json_str = event_schema.Catalog.from_orm(events).json()
+    return json_str
 
 
-def cat_to_dict(obj):
+def cat_to_dict(events: Union[Catalog, Event]) -> dict:
```
Comment (calum-chamberlain) [inline obsplus/events/json.py:31]: Nice change in this function - I'm pretty sure I wouldn't have used an `obj` kwarg, but if you go with the `catalog` kwarg above, can you use `catalog` here as well?

File: obsplus/events/schema.py [hunk 3]
```diff
@@ -0,0 +1,612 @@
+"""
```
Comment (calum-chamberlain) [inline obsplus/events/schema.py:1]: I'm going to assume this is all there given the tests. I haven't checked through this fully though!

File: tests/test_events/test_json.py [hunk 4]
```diff
@@ -73,6 +73,22 @@ def test_catalog_can_be_written(self, test_catalog, json_cat_from_disk):
...
+        event = obspy.read_events()[0]
+        out = cat_to_json(event)
+        assert isinstance(out, str)
+        cat_again = json_to_cat(out)
+        assert len(cat_again) == 1
+
+    def test_list_of_events(self):
+        """Ensure a list of events can be converted to json."""
+        events = obspy.read_events().events
+        out = cat_to_json(events)
```
Comment (calum-chamberlain) [inline tests/test_events/test_json.py:87]: Should the docstring for `cat_to_json` should be `Union[Catalog, Event, Iterable[Event]]`?

**Review by calum-chamberlain — CHANGES_REQUESTED** [review]
Looks good to me - I had a couple of pedantic comments for this pydantic proposal.

File: obsplus/events/schema.py (hunk 3 above)
Comment (d-chambers) [reply obsplus/events/schema.py:1]: Yes, there is at least one test event (#4) which uses every kind of attribute supported by ObsPy. There is a little edge case where the `CompositeTime`'s uncertainties corresponding to int fields don't work (as mentioned above). This is really more of a bug in ObsPy's parsing/typing though and I don't think that it will really effect anyone.

**Comment by d-chambers** [comment]
Thanks @calum-chamberlain, I think I addressed all your concerns, thanks for taking a look at it.

I also added a changelog entry and restructured `cat_to_dict` to use the pydantic model's `dict` method rather than first converting to json which resulted in a minor speedup.

**Comment by d-chambers** [comment]
I also should document here that I timed the json test suite before and after this change. Before using pydantic the test suite took on average 14.5 seconds. After using pydantic the test suite takes about 15 seconds. Considering the huge improvement to the code readability and elimination of recursive functions I think the minor slow-down is worth it.

**Review by calum-chamberlain — APPROVED** [review]
